add option to filter by run number as well as event number

tomeichlersmith opened the pull request:
This is helpful for people like me who have massive samples of heavily-filtered simulations. The best way to generate these samples is across many runs but then merge the resulting files. This results in several runs per file and often an event from one run is aborted by the simulation while the same event number from a different run is the event I wish to resimulate.

tomeichlersmith committed:
add option to filter by run number as well as event number

This is helpful for people like me who have massive samples of
heavily-filtered simulations. The best way to generate these samples is
across many runs but then merge the resulting files. This results in
several runs per file and often an event from one run is aborted by the
simulation while the same event number from a different run is the event
I wish to resimulate.

EinarElen commented on src/SimCore/ReSimulator.cxx:
```diff
@@ -9,23 +9,50 @@ void ReSimulator::configure(framework::config::Parameters& parameters) {
+      if (runs_to_resimulate_.size() == 1) {
+        // single run applied to all events requested
+        should_resim = runs_to_resimulate_.at(0) == run;
+      } else {
+        // more than one run, look for run corresponding to index
+        // of event number found
+        // developer note: this code has a high-liklihood of throwing
+        // the std::out_of_range error if the runs_to_resimulate_ vector
+        // isn't pre-checked to be the same size as events_to_resimulate_
+        should_resim = runs_to_resimulate_.at(
+            std::distance(events_to_resimulate_.begin(), found_event)
```
I think we should catch this error and give the user the context from the developer note in a separate exception when this happens. Always better to see "this thing was wrong" rather than "somewhere a vector was out of range"

EinarElen commented on src/SimCore/ReSimulator.cxx:
```diff
@@ -9,23 +9,50 @@ void ReSimulator::configure(framework::config::Parameters& parameters) {
-    auto found = std::find(std::begin(events_to_resimulate_),
-                           std::end(events_to_resimulate_), eventNumber);
-    if (found == std::end(events_to_resimulate_)) {
+    auto found_event = std::find(std::begin(events_to_resimulate_),
```
I feel like the logic here could be simpler, it is a bit hard to read (at least for me) here, which is why we need so many comments. I think we could create a similar variable to resimulate_all_events that we set in the configuration that tells us what behaviour we are expecting from the run numbers.

EinarElen replied:
If there isn't a straight-forward way to simplify the logic here, I would advocate for creating a vector of event/run pairs during configuration, possibly as a separate function call if it is > a few lines, (encapsulating the messy logic there) and simply do a find here on that vector. In case 2, this vector would just have all the event numbers in events_to_resimulate_ paired with the one run number we care about.

tomeichlersmith replied:
I like creating the pairs in `configure` to make it really clear to future developers how we are interpreting the configuration. As a sanity check, I double checked that `std::find` can work over `std::pair` and it does: https://godbolt.org/z/v4T7a4fKh

EinarElen reviewed (commented):
Looks good (can you tell that my setup is biased towards production that you can run in one big file? :)) 

I do think the logic in produce is a bit complicated, if at all possible I think we should try to simplify it (or at least spread it our/refactor it). As it is, I wouldn't want a new developer to have to read through it

tomeichlersmith committed:
refactor to move config deducing into python

the translation of different format of runs provided by the user into
the actual data structure used during processing is done within python
now so it can be more transparent in error

tomeichlersmith commented:
Ok, I refactored how I implemented it and chose to move the deduction of a configuration from the parameters supplied by the user into the python. I haven't tested running yet, but I wanted to push it up here to see if others think it reads better. @EinarElen ?

EinarElen commented on python/simulator.py.in:
```diff
@@ -6,6 +6,25 @@ with several helpful member functions.
 
 from @PYTHON_PACKAGE_NAME@.Framework.ldmxcfg import Producer
 
+class _EventToReSim:
```
Is that an extra _ ?

tomeichlersmith replied:
It is done on purpose to signal to any users that this is an internal implementation class that is subject to change without warning. A single underscore is just a stylistic convention and is still available to any advanced users who wish to do something special.

EinarElen commented:
This looks reasonable to me, I would just move the check for whether an event should be resimulated into a member function rather than doing it inline with a lambda. i.e. it would read
```C++
if (!resimulate_all) {
  if skip_event(event) {
  abort_event()
} 
}

```

tomeichlersmith committed:
put check on if an event should be skipped into its own function

this is helpful isolation and makes it easier to document how an event
ends up being re-simulated
